akka-stream/src/main/scala/akka/stream/scaladsl/pattern/Debug.scala

```diff
@@ -0,0 +1,58 @@
+/**
+ * Copyright (C) 2014-2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.scaladsl.pattern
+
+import akka.actor.ActorSystem
+import akka.event.{ LogSource, BusLogging, Logging, LoggingAdapter }
+import akka.stream.ActorFlowMaterializer
+import akka.stream.impl.Stages
+import akka.stream.scaladsl.{ FlowOps, Source, FlowGraph, Flow }
+import akka.stream.stage._
+import akka.stream.scaladsl.FlowGraph.Implicits._
+import scala.annotation.implicitNotFound
+import scala.language.implicitConversions
+
+// TODO do we want this to be a "tag on pattern", or put debug() right on Source / Flow?
+object debug {
+
+  implicit def addDebugFlow[In, Out, Mat](f: Flow[In, Out, Mat])(implicit sys: ActorSystem): DebuggableFlow[In, Out, Mat] =
+    new DebuggableFlow(f)
+
+  implicit def addDebugSource[Out, Mat](s: Source[Out, Mat])(implicit sys: ActorSystem): DebuggableSource[Out, Mat] =
+    new DebuggableSource(s)
+
+}
+
+final case class DebuggableFlow[In, Out, Mat](stage: Flow[In, Out, Mat]) extends AnyVal with DebuggableLike[Out, Flow[In, Out, Mat]]
+
+final case class DebuggableSource[Out, Mat](stage: Source[Out, Mat]) extends AnyVal with DebuggableLike[Out, Source[Out, Mat]]
+
+trait DebuggableLike[Out, Repr <: FlowOps[Out, _]] extends Any {
+  protected def stage: Repr
+
+  @implicitNotFound("Cannot find ActorSystem to create LoggingAdapter from") // TODO can we require less? I could use the Materializer yet make it's actor sys akka private?
+  def debug(extract: Out ⇒ Any = identity)(implicit sys: ActorSystem): Repr = { // TODO allow implicit DebugConfig for configuring loglevels?
+    val adapter = akka.event.Logging(sys, /* TODO what here? */ "akka.stream.debug.Debug") // TODO decide on the name + could include materialization name?
```

Comment:
yes, I would be careful with using names with the counters for the LoggingAdapter, because I think slf4j implementations cache all those names in its logger hierarchy (and here it could potentially be many many such names)